### ISSUE TITLE
Support memory request recommendations for jvm based workloads

### ISSUE THREAD

**Issue by vsevel** [kubernetes/autoscaler#5029]
**Which component are you using?**: 

`vertical-pod-autoscaler`

**Is your feature request designed to solve a problem? If so describe the problem this feature should solve.**:

the [GKE documentation](https://cloud.google.com/kubernetes-engine/docs/concepts/verticalpodautoscaler#limitations) indicates: "Vertical Pod autoscaling is not ready for use with JVM-based workloads due to limited visibility into actual memory usage of the workload."

What are those limitations? what visibility do you need exactly? And what use cases does it forbid? It would be nice to have a better understanding of what is supported today (1st step), and (2nd step) develop the missing pieces to have an effective support for JVM based workloads, on-par with other technologies.

Starting in JDK 12 (and backported in JDK 11), [JEP 346: Promptly Return Unused Committed Memory from G1](https://openjdk.org/jeps/346) allows the GC to give back to the OS memory pages that are not necessary anymore. Modern GC all provide a similar feature: https://www.baeldung.com/gc-release-memory

This opens up to memory being used by the process (heap and other) to be more accurate to the real needs of the application.

With this in place, my understanding is that the vpa autoscaler is already accurate to calculate memory requests by watching the consumed memory overtime. Can you confirm? If the calculated recommendation is not accurate, please describe the type of error it is going to introduce: is vpa going to overestimate or underestimate memory requests? If it overestimates memory, this might still be a better solution than to set request=limit (guaranteed).

In my company we have started doing an evaluation of using VPA for memory requests on JVM based workloads (we don't intend to use VPA for memory limits).  On 100 pods, the average configured memory requests is1600Mb (max = 6Gb), and the average vpa recommended memory requests is 1240Mb (max = 4.5Gb). The average saving is 425Mb.  

There are also 20 pods with "memory capped to container limit", which suggests that the memory request recommended by VPA is greater than the limit, which is puzzling.

**Describe the solution you'd like.**:

- describe in documentation the exact limitations when using VPA with memory requests (and limits) for JVM based workloads, depending on the GC being used.
- depending on the above limitations, develop the necessary bits to provide full support for memory recommendations on JVM based workloads.

**Describe any alternative solutions you've considered.**:

None. The other alternative is to do manual tuning and setup of requests in the pod spec.

**Additional context.**:

see this [kubecon talk](https://kccnceu2022.sched.com/event/ytkd/how-lombard-odier-deployed-vpa-to-increase-resource-usage-efficiency-vincent-sevel-lombard-odier-sa)

cc @mwielgus @matthyx

**Comment by jbartosik** [comment]
I didn't look into that but from what I recall with talking to others VPA recommendations might explode for JVM workloads.

It will run garbage collection when it thinks it's close to running out of memory. So if you're looking at memory usage you'll see that:

* memory usage is growing slowly (until it hits threshold for garbage collection), then
*  it goes down rapidly (when GC runs), then
* it goes back to growing slowly.

VPA looks at daily peak of memory usage (max memory usage in 24h window for each container). So if the memory usage cycle takes less than 24h to VPA this will look like the container is using ${GC threshold}.

${GC threshold} should depend on memory request (otherwise increasing memory request doesn't make sense). So if VPA grows memory the threshold will also increase. But since memory usage (as observed by VPA) depends on the threshold memory usage (observed by VPA) will also grow, so recommendation will grow and so on until VPA starts capping recommendation because it wouldn't fit in any node in the cluster.

**Comment by jbartosik** [comment]
> In my company we have started doing an evaluation of using VPA for memory requests on JVM based workloads (we don't intend to use VPA for memory limits). On 100 pods, the average configured memory requests is1600Mb (max = 6Gb), and the average vpa recommended memory requests is 1240Mb (max = 4.5Gb). The average saving is 425Mb.

This sound like memory usage is not growing anywhere close to the request (which conflict with an assumption in my previous comment). Can you share why?

**Comment by vsevel** [comment]
just to be sure, `${GC threshold}` is the value at which the gc is going to kick in to do a full gc?
actually, the gc is a bit more complex, and it depends on which gc impl we are talking about. there could be different behaviors depending on using `g1` vs `zgc` for instance.
but usually there are zones (e.g. new and old). when created the objects stay in the `new` zone for a while. if they resist several gc cycles, they will eventually land in the `old` zones where another algorithm takes place.
99% of objects get gced in the `new` zone. that is where you have the real app activity.
I have a memory intensive app (lucene indexer server), which has a 0 activity (0 ms per minute, 0 cycles per minute) on the `G1 Old Generation`, and in average around 400 ms per minute (min 200ms, max 1100 ms) of activity for the `G1 Young Generation`.
so the **heap** memory profile (old+new) for this app is fairly stable. it does increase slowly like you are saying, but this will probably take days before there is a full gc. 
now it is my responsibility to set a `Xmx`, which is big enough to accommodate my use cases, and tight enough that I will not waste committed heap memory that the jvm does not feel like returning. if this is the case. g1 has made a lot of effort to not waste memory, like the JEP 346 I was referring to. I need to watch this.
but regardless, the heap is not the only memory space to account for in a jvm. there is all the memory consumed by the process that is not part of the heap.
the extra memory consumed by the jvm can be 50 to 100% of the heap. there are jvm flags for this: `MaxRAMFraction=2` means `1/2` of max memory in the container will be reserved for your heap. so if your container has a limit of 1Gb, your heap (Xmx) will be 512Mb.
The point is that it might be difficult for VPA to assess correctly the heap, but recommendations are still useful for everything else. and that is not necessarily tiny.

> This sound like memory usage is not growing anywhere close to the request (which conflict with an assumption in my previous comment). Can you share why?

with that in mind, I can imagine a few things that happen there, and appear to be working:
- the heap is regularly gced, so the value looks roughly like what is actually needed
- the non heap memory is the one going up and down

worst case scenario I am thinking that even if VPA is not assessing correctly the heap (because it contains objects that could be gced), the value that VPA captures (so total memory of the process) is a value that may be oversized. but can't be undersized (otherwise the app would have crashed). so it is somewhere between correct and oversized. and it will be very much oversized if the Xmx is way too big for the app (too much comfort) and the gc is lazy in the new zone (objects stay a long time in the heap before being gced).

so I am thinking VPA should be relevant even for memory requests on jvm workloads assuming:
- we use a gc that does much of its work in the `new` zone
- the heap (Xmx) is appropriately sized (not too small, not too big)

there are situations however, I do not understand how VPA could calculate a value that big.
for instance I have this deployment:
```
      containers:
        - resources:
            limits:
              cpu: '2'
              ephemeral-storage: 1Gi
              memory: 728Mi
            requests:
              cpu: 100m
              ephemeral-storage: 1Gi
              memory: 728Mi
...
            - name: JAVA_OPTS_APPEND
              value: '-Xms128m -Xmx256m'
```

and VPA calculated:
```
  recommendation:
    containerRecommendations:
      - containerName: container-efxservices
        lowerBound:
          cpu: 25m
          memory: '857515809'
        target:
          cpu: 93m
          memory: '1102117711' ==> 1051 Mb??
        uncappedTarget:
          cpu: 93m
          memory: '1102117711' ==> 1051 Mb??
        upperBound:
          cpu: 743m
          memory: '6500245275' ==> 6199 Mb???
```

how could it recommend a value greater than the limit? this can't be something it saw was consumed...
on this app I see the heap old+new going regularly between 120Mb and 185Mb every 1 or 2 minutes, and the gc activity is low (time spent and number of cycles, and the container is stable around 750 Mb (it is not the g1 in that case, so unused memory is not provided back to the os). the app has gone through 2 restarts (not OOM killed I believe) in the last few hours. so there might be a reason for the VPA calculation?
regardless, I do not understand how you would come up with values greater than the limit.

the bottom line is that even if heap behavior is a challenge in a jvm workload, worst case you might tend to overestimate the heap needed and you should be relevant for the non heap memory.
do you think that makes sense?

**Comment by vsevel** [comment]
have you been able to give it some thoughts?
cc @matthyx

**Comment by satyamsundaram** [comment]
What is the progress on this? Is VPA still not recommended for JVM-based apps?
